README
DBD::SQLAnywhere -- a SQLAnywhere interface for Perl 5.

   Portions Copyright (c) 1994,1995,1996            Tim Bunce
   Portions Copyright (c) 1996-2010		    iAnywhere Solutions, Inc.

   For license information, please see license.txt included in this
   distribution.

*BEFORE* BUILDING, TESTING AND INSTALLING this you will need to:

    Build, test and install Perl 5 (at least 5.6.0 is recommended).
    It is very important to test it and install it!

    Build, test and install the DBI module (at least DBI 1.51 is
    required) if it is not already included in your perl distribution.

    No SQL Anywhere software needs to be installed to build this
    module; however, a SQL Anywhere client installation is required in
    order to use the module.

    A full SQLAnywhere client+server installation is recommended for
    testing this module otherwise the tests will be skipped.

Building:
    On UNIX:
	[ Linux, Solaris, AIX, HP-UX, etc. ]

	source sa_config.sh (or sa_config.csh) from the SQL Anywhere
	installation

	perl Makefile.PL
	make
	copy demo.db from the SQL Anywhere installation directory
	    to the current directory
	make test
	make install
    
    On Win32:
	[ requires Microsoft C and ActiveState's ActivePerl
	  (www.activestate.com) ]

	NOTE: ActivePerl 5.6.0 build 616 or later is required.

	perl Makefile.PL
	nmake
	copy demo.db from the SQL Anywhere installation directory
	    to the current directory
	nmake test
	nmake install

    Do not hand-edit the generated Makefile as those changes will be
    lost the next time 'perl Makefile.PL' is invoked. Always try to
    make changes via the Makefile.PL command line and/or editing
    Makefile.PL.

---------------------------------------------------------------

Notes:

For information on how to use DBD::SQLAnywhere, please invoke the
following after DBD::SQLAnywhere has been installed:

    perldoc DBD::SQLAnywhere

Prior to installation, you may instead run the following when
SQLAnywhere.pm is in your current directory:

    perldoc SQLAnywhere.pm

Examples can be found in the 'eg' directory:
    connect.pl  -> Demonstrates a simple connect.
    retrieve.pl -> Demonstrates retrieving data from a table and displaying the result set.
    blobs.pl    -> Demonstrates the use of blobs

For further examples, consult the test scripts located in the 't'
directory.



Changes
Changes in DBD::SQLAnywhere 2.08 23rd November 2010, 10th January 2011
  Prepare for removal of PERL_POLLUTE support.
  Moved usage documentation into SQLAnywhere.pm as POD documentation. README
    only contains information for building. CPAN search "all" function actually
    only looks at the POD documentation.
  table_info, etc now use SYSUSER rather than SYSUSERPERM. SYSUSERPERM was
    deprecated in v10 and can only be read by DBA.

Changes in DBD::SQLAnywhere 2.07 22nd October 2010
  Somehow, the v2.06 package that I uploaded to PAUSE contained the old tests.
    I'm just going to try again but that requires a new version number.
  I also added a dependency on Test::Simple (which contains Test::More)

Changes in DBD::SQLAnywhere 2.06 21st October 2010
  Removed leak of bind_names hv for hostvars
  Added PERL_NO_GET_CONTEXT for efficiency on threaded perl.
  Changed tests to pass/skip if SQL Anywhere is not installed or demo.db
    is not accessible.

Changes in DBD::SQLAnywhere 2.05 19th October 2010
  Fixed resource cleanup on failed connect
  Added META.yml

Changes in DBD::SQLAnywhere 2.04 19th October 2010
  Used refcounting for SACAPI context since DBI can free driver handles
    before all database handles have been destroyed.

Changes in DBD::SQLAnywhere 2.03 4th December 2009
  Fixed fetching of double/floating point values
  Updated sacapi files.

Changes in DBD::SQLAnywhere 2.02 1st October 2009
  Previous driver (version 2.01) broke support to SQLAnywhere external through Perl. This
    has been fixed.

Changes in DBD::SQLAnywhere 2.01 2nd June 2009
  Improved threading support with SACAPI v1 and added SACAPI v2 support.
  Removed primary_key func (to rely on the one in DBI.PM)
  Added statistics_info (returns NULL) and last_insert_rowid

Changes in DBD::SQLAnywhere 2.00 9th May 2008
  Changed to use simplified SQLAnywhere C API (SACAPI). The driver can now be built
    without having SQLAnywhere installed but SQLAnywhere is required to use the driver.
  Added support for the the get_info method.

Changes in DBD::SQLAnywhere 1.16 2nd Nov 2007, 28th Feb 2008, 17th March 17 2008
  (version 1.15 was not released)
  Fixed installation location & increased version number so that we can detect it
  Support new bin32/bin64 directories for SQLAnywhere 11.0 on Windows
  Automated CPAN testing requires a successful exit from Makefile.PL if the module
    cannot be built (due to libs not found). Go figure.

Changes in DBD::SQLAnywhere 1.15 17th July 2007
  (version 1.14 was not released)
  Improved table_info routine
  Added column_info, primary_key, primary_key_info routines

Changes in DBD::SQLAnywhere 1.14 18th Jan 2007
  (previous changes to 1.14 were not released)
  Fixed UNIX builds. With the SQLAnywhere 10.0 release the bin and lib
  directories in the UNIX install have been renamed to bin32/bin64 and
  lib32/lib64. Fixed Makefile.PL to search for libraries in the correct path.

Changes in DBD::ASAny 1.14 21th Nov 2006
  (previous changes to 1.14 were not released)
  Fixed Windows x64 builds.

Changes in DBD::ASAny 1.14 7th Sept 2006
  (previous changes to 1.14 were not released)
  Added a manifest to the DLL for VS2005 builds.

Changes in DBD::SQLAnywhere 1.14 9th Dec 2005
  (previous changes to 1.14 were not released)
  Renamed from DBD::ASAny to DBD::SQLAnywhere for Jasper release
    which includes changes to use demo.db instead of asatest.db.
  Added -D_CRT_SECURE_NO_DEPRECATE to avoid VS2005 warnings.
  Cleaned up the README & the examples a bit.
  Added README.NT with CRLF line endings.

Changes in DBD::ASAny 1.14  29th Jan 2004, 5th Apr 2005
  Parsing for placeholders assumed that only single-quotes could be used to 
    delimit a string. Now, either single or double quotes can be used.
  If fetch was called when there were no more rows, the driver set the err 
    value to (integer) zero but DBI could interpret that as a warning because
    it converted it to a string and saw that it was non-empty. We now set
    err to the empty string.

Changes in DBD::ASAny 1.13	7th Mar 2002, 3rd Sept 2003, 29th Oct 2003
  Support 64bit UNIX platforms
  Avoid compile-time warning about assign_blob()
  Added /opt/sybase/SYBS* as default installation directories to search for ASA
  An error was reported if execute() was called twice on the same statement
    handle. Changed execute() to close a previously opened cursor using that
    statement. Also, the prepared statement was being dropped during
    finish() rather during destroy & that prevented the prepare_cache from
    working.
  Updated README to warn about resource governor errors now that statement
    handles are not dropped until destroy time.
  Set the "state" value on the handle when an error occurs.
  Fetching 64-bit integers failed when they contained more than 8 digits
    because we fetch them as strings but left the length as the size of the
    bigint datatype (8 bytes).
  Fixed additional compiler warnings from VC7.
  Added CLONE method.
  Added better handling for detecting new ASA versions -- should handle
   up to version 15 now (version 9 is the highest shipping version).
  Implemented the blob_read function which is an undocumented DBI API.
  Changed the long string/binary fetching code so that long character strings
    are fetched as varchar rather than binary so that character set translation
    will be applied to the string.
  Added a commit operation after a cursor is closed if AutoCommit is enabled.
    Previously, a commit was only performed if a non-cursor statement was
    executed. DML executed in procedures returning result sets were not
    committed.
  Added "-r" switch to sqlpp to generate thread-safe code.
  Added dbtasks library to UNIX link lines.
  Link against "_r" reentrant versions of ASA libraries when using ithreads
    on UNIX platforms.
  Avoid warning when building with VS.NET 2003 which has deprecated the
    /Gf switch (automatically added by MakeMaker). Changed it to /GF.

SQLAnywhere.pm
#====================================================
#
#      Copyright 2008-2010 iAnywhere Solutions, Inc.
#
#   Licensed under the Apache License, Version 2.0 (the "License");
#   you may not use this file except in compliance with the License.
#   You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS,
#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#
#   See the License for the specific language governing permissions and
#   limitations under the License.
#
#   While not a requirement of the license, if you do modify this file, we
#   would appreciate hearing about it. Please email
#   sqlany_interfaces@sybase.com
#
#====================================================
require 5.002;
use strict;
use warnings;
{
    package DBD::SQLAnywhere;

    use DBI ();
    use DynaLoader ();
    use Exporter ();

    our $VERSION = '2.08';
    our @ISA = qw(DynaLoader Exporter);
    our %EXPORT_TAGS = (
	asa_types => [ qw(
	    ASA_SMALLINT ASA_INT ASA_DECIMAL ASA_FLOAT ASA_DOUBLE ASA_DATE
	    ASA_STRING ASA_FIXCHAR ASA_VARCHAR ASA_LONGVARCHAR ASA_TIME
	    ASA_TIMESTAMP ASA_TIMESTAMP_STRUCT ASA_BINARY ASA_LONGBINARY
	    ASA_VARIABLE ASA_TINYINT ASA_BIGINT ASA_UNSINT ASA_UNSSMALLINT
	    ASA_UNSBIGINT ASA_BIT ) ],
    );
    Exporter::export_ok_tags( 'asa_types' );

    my $Revision = substr(q$Revision: 1.57 $, 10);

    require_version DBI 1.51;

    bootstrap DBD::SQLAnywhere $VERSION;

    our $err = 0;		# holds error code   for DBI::err    (XXX SHARED!)
    our $errstr = "";	# holds error string for DBI::errstr (XXX SHARED!)
    our $drh = undef;	# holds driver handle once initialised

    sub CLONE {
	$drh = undef;
    }

    sub driver {
	return $drh if $drh;
	my($class, $attr) = @_;

	$class .= "::dr";

	# not a 'my' since we use it above to prevent multiple drivers

	$drh = DBI::_new_drh($class, {
	    'Name' => 'SQLAnywhere',
	    'Version' => $VERSION,
	    'Err'    => \$DBD::SQLAnywhere::err,
	    'Errstr' => \$DBD::SQLAnywhere::errstr,
	    'Attribution' => 'SQLAnywhere DBD by John Smirnios',
	    });

	if( !DBD::SQLAnywhere::dr::driver_init( $drh ) ) {
	    undef( $drh );
	}

	$drh;
    }

    1;
}


{   package DBD::SQLAnywhere::dr; # ====== DRIVER ======
    use strict;

    sub connect {
	my($drh, $dbname, $user, $auth, $attr)= @_;

	# NOTE!
	# 
	# For SQLAnywhere, $dbname and $user are appended to form an
	# SQLAnywhere connection string. 'UID=' is prefixed onto $user
	# if necessary. If $auth is nonempty, 'PWD=' is prefixed.
	# If dbname starts with something that doesn't look like
	# a connect string parameter ('label=value;' format) then
	# 'ENG=' is prefixed.
	my $conn_str;
	my $sqlcap;

	if( defined( $dbname ) ) {
	    $conn_str = $dbname;
	    $conn_str =~ s/^[\s;]*//;
	    $conn_str =~ s/[\s;]*$//;
	    if( $conn_str =~ /^[^=;]+($|;)/ ) {
		$conn_str = 'ENG=' . $conn_str;
	    }
	    
	    # look for undocumented option indicating the sqlca to use for
	    # server-side perl
	    if( $conn_str =~ /^ENG=saperl;sa_perl_sqlca=(0x)?([0-9a-fA-F]*)$/ ) {
	    	$sqlcap = $2;
	    }
	} else {
	    $conn_str = '';
	}
	if( defined( $user ) && ($user ne '') ) {
	    if( $user =~ /=/ ) {
		$conn_str .= ';' . $user;
	    } else {
		$conn_str .= ';UID=' . $user;
	    }
	}
	if( defined( $auth ) && ($auth ne '') ) {
	    $conn_str .= ';PWD=' . $auth;
	}

	# create a 'blank' dbh
	my $dbh = DBI::_new_dbh($drh, {
	    'Name' => $conn_str,
	    'USER' => $user, 'CURRENT_USER' => $user,
	    });

	# Call SQLAnywhere connect func in SQLAnywhere.xs file
	# and populate internal handle data.

	if( !DBD::SQLAnywhere::db::_login($dbh, $conn_str, 
					  (defined $sqlcap) ? $sqlcap : '', '', $attr) ) {
	    return undef;
	}

	$dbh;
    }
}


{   package DBD::SQLAnywhere::db; # ====== DATABASE ======
    use strict;

    sub prepare {
	my($dbh, $statement, @attribs)= @_;

	# create a 'blank' sth

	my $sth = DBI::_new_sth($dbh, {
	    'Statement' => $statement,
	    });

	# Call SQLAnywhere OCI oparse func in SQLAnywhere.xs file.
	# (This will actually also call oopen for you.)
	# and populate internal handle data.

	DBD::SQLAnywhere::st::_prepare($sth, $statement, @attribs)
	    or return undef;

	$sth;
    }


    sub ping {
	my($dbh) = @_;
	# we know that DBD::SQLAnywhere prepare does a describe so this will
	# actually talk to the server and is a valid and cheap test.
	return 1 if $dbh->prepare("select 1");
	return 0;
    }


# Use the DBI-provided quote routine
#    sub quote {
#	my($dbh, $value) = @_;
#	return $value;
#    }


#    sub quote_identifier {
#	my($dbh, $name) = @_;
#	return "\"".$name."\"";
#    }


    sub table_info {
	my($dbh,$catalogue,$schema,$table,$type)       = @_;		# XXX add qualification

	if ( !defined($schema) || $schema eq "" ) {
	    $schema = '%';
	}

	if ( !defined($table) || $table eq "" ) {
	    $table = '%';
	}

	if ( !defined($type) || $type eq "" ) {
	    # $type = 'TABLE,VIEW,SYSTEM TABLE,GLOBAL TEMPORARY,LOCAL TEMPORARY,ALIAS,SYNONYM';
	    $type = '%';
	}

	my $sth = $dbh->prepare("
        select
	    NULL as TABLE_CAT,
	    u.user_name as TABLE_SCHEM,
	    t.table_name as TABLE_NAME,
	    (if t.table_type = 'BASE' then (if t.creator = 0 then 'SYSTEM ' else '' endif) ||'TABLE'
		else (if t.table_type = 'GBL TEMP' then 'GLOBAL TEMPORARY' 
		      else t.table_type
		      endif)
		endif) as TABLE_TYPE,
	    t.remarks as REMARKS
	from SYS.SYSTABLE t, SYS.SYSUSER u
	where t.creator = u.user_id
	  and u.user_name  like ?
 	  and t.table_name like ?
	  and TABLE_TYPE   like ?
	order by u.user_name, t.table_name
	") or return undef;
# and TABLE_TYPE  IN (?)
	$sth->bind_param( 1, $schema );
	$sth->bind_param( 2, $table );
	$sth->bind_param( 3, $type );
	$sth->execute or return undef;
	$sth;
    }


    sub type_info_all {
	my ($dbh) = @_;
	my $names = {
	    TYPE_NAME		=> 0,
	    DATA_TYPE		=> 1,
	    COLUMN_SIZE		=> 2,
	    LITERAL_PREFIX	=> 3,
	    LITERAL_SUFFIX	=> 4,
	    CREATE_PARAMS	=> 5,
	    NULLABLE		=> 6,
	    CASE_SENSITIVE	=> 7,
	    SEARCHABLE		=> 8,

	    UNSIGNED_ATTRIBUTE	=> 9,
	    FIXED_PREC_SCALE	=>10,
	    AUTO_UNIQUE_VALUE	=>11,
	    LOCAL_TYPE_NAME	=>12,
	    MINIMUM_SCALE	=>13,
	    MAXIMUM_SCALE	=>14,
	    SQL_DATA_TYPE	=>15,
	    SQL_DATETIME_SUB	=>16,
	    NUM_PREC_RADIX	=>17,
	};
	my $ti = [
	  $names,
	    [ 'bit', -7, 1, undef, undef, undef, 1, 0, 3,
	      1, undef, 0, undef, undef, undef, -7, undef, undef
	    ],
	    [ 'tinyint', -6, 4, undef, undef, undef, 1, 0, 3,
	      0, undef, 0, undef, undef, undef, -6, undef, undef
	    ],
	    [ 'bigint', -5, 20, undef, undef, undef, 1, 0, 3,
	      0, undef, 0, undef, undef, undef, -5, undef, undef
	    ],
	    [ 'unsigned bigint', -5, 20, undef, undef, undef, 1, 0, 3,
	      1, undef, 0, undef, undef, undef, -5, undef, undef
	    ],
	    [ 'long binary', -4, 2147483647, '\'', '\'', undef, 1, 0, 3,
	      undef, undef, undef, undef, undef, undef, -4, undef, undef
	    ],
	    [ 'binary', -2, 65535, '\'', '\'', 'max length', 1, 0, 3,
	      undef, undef, undef, undef, undef, undef, -2, undef, undef
	    ],
	    [ 'varbinary', -2, 65535, '\'', '\'', 'max length', 1, 0, 3,
	      undef, undef, undef, undef, undef, undef, -2, undef, undef
	    ],
	    [ 'long varchar', -1, 2147483647, '\'', '\'', undef, 1, 0, 3,
	      undef, undef, undef, undef, undef, undef, -1, undef, undef
	    ],
	    [ 'char', 1, 65535, '\'', '\'', 'max length', 1, 0, 3,
	      undef, undef, undef, undef, undef, undef, 1, undef, undef
	    ],
	    [ 'decimal', 2, 127, undef, undef, 'precision, scale', 1, 0, 3,
	      0, 0, 0, undef, 0, 127, 2, undef, 10
	    ],
	    [ 'numeric', 2, 127, undef, undef, 'precision, scale', 1, 0, 3,
	      0, 0, 0, undef, 0, 127, 2, undef, 10
	    ],
	    [ 'money', 3, 4, undef, undef, undef, 1, 0, 3,
	      0, 1, 0, undef, 4, 4, 3, undef, 10
	    ],
	    [ 'smallmoney', 3, 4, undef, undef, undef, 1, 0, 3,
	      0, 1, 0, undef, 4, 4, 3, undef, 10
	    ],
	    [ 'integer', 4, 10, undef, undef, undef, 1, 0, 3,
	      0, undef, 0, undef, 0, 0, 4, undef, undef
	    ],
	    [ 'unsigned int', 4, 10, undef, undef, undef, 1, 0, 3,
	      1, undef, 0, undef, undef, undef, 4, undef, undef
	    ],
	    [ 'smallint', 5, 6, undef, undef, undef, 1, 0, 3,
	      0, undef, 0, undef, 0, 0, 5, undef, undef
	    ],
	    [ 'unsigned smallint', 5, 5, undef, undef, undef, 1, 0, 3,
	      1, undef, 0, undef, undef, undef, 5, undef, undef
	    ],
	    [ 'double', 6, 64, undef, undef, undef, 1, 0, 3,
	      0, undef, 0, undef, undef, undef, 6, undef, 2
	    ],
	    [ 'float', 7, undef, undef, undef, undef, 1, 0, 3,
	      0, undef, 0, undef, undef, undef, 7, undef, 32
	    ],
	    [ 'double', 8, 64, undef, undef, undef, 1, 0, 3,
	      0, undef, 0, undef, undef, undef, 8, undef, 2
	    ],
	    [ 'varchar', 12, 65535, '\'', '\'', 'max length', 1, 0, 3,
	      undef, undef, undef, undef, undef, undef, 12, undef, undef
	    ]
        ];
	return $ti;
    }


    sub column_info {
	my($dbh,$catalogue,$schema,$table,$column)       = @_;		# XXX add qualification

	if( !defined($schema) || $schema eq "" ) {
	    $schema = '%';
	}

	if( !defined($table) || $table eq "" ) {
	    $table = '%';
	}

	if( !defined($column) || $column eq "" ) {
	    $column = '%';
	}

	my $sth = $dbh->prepare("
        select
	    NULL as TABLE_CAT,
	    u.user_name as TABLE_SCHEM,
	    t.table_name as TABLE_NAME,
	    c.column_name as COLUMN_NAME,
	    d.domain_id as DATA_TYPE, 
            d.domain_name AS TYPE_NAME,
	    c.width AS COLUMN_SIZE,
	    c.width AS BUFFER_LENGTH,
            c.width AS DECIMAL_DIGITS, 
	    c.scale AS NUM_PREC_RADIX,
	    IF c.nulls = 'Y' THEN 1 ELSE 0 ENDIF AS NULLABLE,
	    c.remarks AS REMARKS,
            c.\"default\" AS COLUMN_DEF, 
            d.domain_name AS SQL_DATA_TYPE,
            NULL AS SQL_DATETIME_SUB,
	    c.width AS CHAR_OCTET_LENGTH,
	    c.column_id AS ORDINAL_POSITION,
	    c.nulls AS IS_NULLABLE,
            NULL AS CHAR_SET_CAT,
	    NULL AS CHAR_SET_SCHEM,
	    NULL AS CHAR_SET_NAME,
	    NULL AS COLLATION_CAT,
	    NULL AS COLLATION_SCHEM,
	    NULL AS COLLATION_NAME,
	    NULL AS UDT_CAT,
	    NULL AS UDT_SCHEM,
	    NULL AS UDT_NAME,
	    NULL AS DOMAIN_CAT,
	    NULL AS DOMAIN_SCHEM,
	    NULL AS DOMAIN_NAME,
	    NULL AS SCOPE_CAT,
	    NULL AS SCOPE_SCHEM,
	    NULL AS SCOPE_NAME,
	    NULL AS MAX_CARDINALITY,
	    NULL AS DTD_IDENTIFIER,
	    NULL AS IS_SELF_REF
	    from SYS.SYSTABLE t
	   , SYS.SYSUSER u
	   , SYS.SYSCOLUMN c
	   , SYS.SYSDOMAIN d
	where t.creator     = u.user_id
	  and t.table_id    = c.table_id
	  and c.domain_id   = d.domain_id
	  and u.user_name   like ?
	  and t.table_name  like ?
	  and c.column_name like ?
	order by c.column_id
	") or return undef;
	$sth->bind_param(1, $schema); 
	$sth->bind_param(2, $table); 
	$sth->bind_param(3, $column); 
	$sth->execute or return undef;
	$sth;
    }


    sub primary_key_info {
	my($dbh,$catalogue,$schema,$table,$column)       = @_;		# XXX add qualification

	if ( !defined($schema) || $schema eq "" ) {
	    $schema = '%';
	}

	if ( !defined($table) || $table eq "" ) {
	    $table = '%';
	}

	if ( !defined($column) || $column eq "" ) {
	    $column = '%';
	}

	my $sth = $dbh->prepare("
        select
	    NULL as TABLE_CAT,
	    u.user_name as TABLE_SCHEM,
	    t.table_name as TABLE_NAME,
	    c.column_name as COLUMN_NAME,
	    c.column_id AS KEY_SEQ,
	    i.index_name as PK_NAME
	from SYS.SYSTABLE t
       , SYS.SYSUSER u
       , SYS.SYSCOLUMN c
       , SYS.SYSIDX i
	where t.creator     = u.user_id
	  and t.table_id    = c.table_id
	  and t.table_id    = i.table_id
	  and i.index_id    = 0
      and c.pkey        = 'Y'
      and u.user_name   like ?
      and t.table_name  like ?
      and c.column_name like ?
    order by c.column_id
	") or return undef;
	$sth->bind_param(1, $schema); 
	$sth->bind_param(2, $table); 
	$sth->bind_param(3, $column); 
	$sth->execute or return undef;
	$sth;
    }

    sub get_info {
        my($dbh, $info_type) = @_;
        require DBD::SQLAnywhere::GetInfo;
        my $v = $DBD::SQLAnywhere::GetInfo::info{int($info_type)};
        $v = $v->($dbh) if ref $v eq 'CODE';
        return $v;
    }

    sub statistics_info {
	my($dbh,$catalogue,$schema,$table,$unique_only,$quick) = @_;	# XXX add qualification

	if ( !defined($schema) || $schema eq "" ) {
	    $schema = '%';
	}

	if ( !defined($table) || $table eq "" ) {
	    $table = '%';
	}

	if ( defined($unique_only) && $unique_only == 1 ) {
	    $unique_only = 2;
	}

        # quick ignored for now
	if ( !defined($quick) || $quick eq "" || $quick != 1 ) {
	    $quick = 0;
	}

	my $sth = $dbh->prepare("
        select
	    NULL as TABLE_CAT,
	    u.user_name as TABLE_SCHEM,
	    t.table_name as TABLE_NAME,
	    IF i.\"unique\" = '1' THEN 1 ELSE 0 ENDIF as NON_UNIQUE,
	    t.table_name ||'.'|| i.index_name as INDEX_QUALIFIER,
	    i.index_name as INDEX_NAME,
	    'table' as TYPE,
	    NULL as ORDINAL_POSITION,
	    NULL as COLUMN_NAME,
	    NULL as ASC_OR_DESC,
	    NULL as CARDINALITY,
	    NULL as PAGES,
	    NULL as FILTER_CONDITION
	from SYS.SYSTABLE t
	, SYS.SYSUSER u
	, SYS.SYSIDX i
	where t.creator  = u.user_id
	  and t.table_id = i.table_id
	  and u.user_name  like ?
 	  and t.table_name like ?
 	  and i.\"unique\"     like ?
	order by u.user_name, t.table_name
	") or return undef;
	$sth->bind_param( 1, $schema );
	$sth->bind_param( 2, $table );
	$sth->bind_param( 3, $unique_only );
#	$sth->bind_param( 4, $quick );
	$sth->execute or return undef;
	$sth;
    }

    sub last_insert_rowid {
	my($dbh,$source,$col) = @_;

	my $sth = $dbh->prepare("
         select \@\@IDENTITY
	") or return undef;
	$sth->execute or return undef;
        my @ida = $sth->fetchrow_array();
        return @ida ? $ida[0] : undef;
    }


}   # end of package DBD::SQLAnywhere::db


{   package DBD::SQLAnywhere::st; # ====== STATEMENT ======

    # all done in XS
}

1;

__END__

=head1 NAME

DBD::SQLAnywhere - SQLAnywhere database driver for DBI

=head1 SYNOPSIS

  use DBI;

  $dbh = DBI->connect( "dbi:SQLAnywhere:ENG=demo;UID=$userid;PWD=$passwd", '', '' );

  $dbh = DBI->connect( 'dbi:SQLAnywhere:ENG=demo', $userid, $passwd );

  # Use 'perldoc DBI' for detailed information about DBI.

=head1 DESCRIPTION

DBD::SQLAnywhere is a Perl database driver (DBD) module that works
with the L<DBI> module to provide access to Sybase SQL Anywhere
databases.

=head2 Connecting to SQL Anywhere

If you are not already familiar with SQL Anywhere connection
parameters, please refer to the SQL Anywhere documentation.

SQL Anywhere connection parameters can be passed to DBD::SQLAnywhere
by placing the list of parameters after 'dbi:SQLAnywhere:' in the
first parameter to connect(). The connection parameters are specified
as a list of LABEL=value pairs that are delimited by semicolons.

Example:

    $dbh = DBI->connect( 'dbi:SQLAnywhere:ENG=demo;UID=dba;PWD=sql', '', '' );

If the second argument to connect() is nonblank, it is assumed to be a
user name and UID=argument2 will be appended to the SQL Anywhere
connection string.

Similarly, if the third argument is nonblank, it is assumed to be a
password and PWD=argument3 will be appended to the SQL Anywhere
connection string.

The following is equivalent to the example above:

    $dbh = DBI->connect( 'dbi:SQLAnywhere:ENG=demo', 'dba', 'sql' );

=head2 Prepared Statement and Cursor Limits

To help detect handle leaks in client applications, SQL Anywhere
defaults to limiting the number of prepared statements and open
cursors that any connection can hold at one time to 50 of each. If
that limit is exceeded, a "Resource governor ... exceeded" error is
reported. If you encounter this error, make sure you are dropping all
of your statement handles and, if so, consult the SQL Anywhere
documentation for the MAX_CURSOR_COUNT and MAX_STATEMENT_COUNT
options.

Note that prepared statements are not dropped from the SQL Anywhere
server until the statement handle is destroyed in the perl
script. Calling finish() is not sufficient to drop the handle that the
server is holding onto: use "undef" instead or reuse the same perl
variable for another handle.

Be careful when using prepare_cached() since the cache will
hold onto statement handles.

=head1 REQUIREMENTS

As of version 2.0, DBD::SQLAnywhere can be built (but not used)
without SQL Anywhere installed. To use DBD::SQLAnywhere, an
installation of the SQL Anywhere client software is required and must
include the "dbcapi" component (dbcapi.dll on Windows and
libdbcapi.so/libdbcapi_r.so on UNIX). Dbcapi is included in the client
installation of SQL Anywhere 11.0.0 and later. To use SQL Anywhere
version 10.0.0 or 10.0.1 with this DBD driver, download and
install a current EBF for your version of SQL Anywhere from
www.sybase.com.

=head1 DEPENDENCIES

L<DBI>

L<Test::Simple>

=head1 AUTHOR

John Smirnios (smirnios@sybase.com).

Based on a driver written by Tim Bunce.

=head1 COPYRIGHT

Portions Copyright (c) 1994,1995,1996           Tim Bunce
Portions Copyright (c) 1996-2010		iAnywhere Solutions, Inc.

For license information, please see license.txt included in this
distribution.

=cut
